1.error-handling/index.js

```diff
@@ -1,19 +1,52 @@
 class MissingPropertyError extends Error {
   // please update the class
+  constructor(message) {
+    super(message);
+    this.name = 'MissingPropertyError';
+  }
 }
 
 class InvalidTypeError extends Error {
   // please update the class
+  constructor(message) {
+    super(message);
+    this.name = 'InvalidTypeError';
+  }
 }
 
 function parseAndValidateJSON(jsonString) {
   try {
     let parsedData = JSON.parse(jsonString);
     // use if statements and throw the errors
+    if (!parsedData.name) {
+      throw new MissingPropertyError(`Property "name" is missing`);
+    }
+    
+    if (!parsedData.age) {
+      throw new MissingPropertyError(`Property "age" is missing`);
+    }
+    
+    if (typeof parsedData.name !== 'string') {
+      throw new InvalidTypeError(`Property "name" should be of type "string"`);
+    }
+
+    if (typeof parsedData.age !== 'number') {
+      throw new InvalidTypeError(`Property "age" should be of type "number"`);
+    }
 
     return parsedData;
-  } catch (error) {
+  } catch (err) {
     // handle the error here
+    const { name, message } = err;
+    if (err instanceof MissingPropertyError) {
+      console.log(`${name}: ${message}`);
+    } else if (err instanceof InvalidTypeError) {
+      console.log(`${name}: ${message}`);
+    } else if (err instanceof SyntaxError) {
+      console.log(`${name}: ${message}`);
+    } else {
+      console.log(`${name}: ${message}`);
+    }
```

Comment:
apakah penggunaan instanceof seperti ini? agak berasa redundant ya

Reply:
Seperti ini:

```
if (error instanceof MissingPropertyError || error instanceof InvalidTypeError || error instanceof SyntaxError) {
      console.log(`${error.name}: ${error.message}`);
    } else {
      throw error;  // throw kembali unknown errors
    }
```

Reply:
lalu kita buat catch lagi di bawahnya? untuk menangkap error yg kita throw dari else itu?

Reply:
Tidak perlu. Ketika kita menemui situation dimana kita tidak tau bagaimana cara menghandlenya, kita harus throw kembali error tersebut, ini dinamakan practice untuk `avoid silent failures`.

Contohnya, programmer melakukan typo seperti ini:
```
funcUtilize(parsdeData)
```

Hal seperti itu daripada di handle justru lebih baik kalau menghentikan eksekusi supaya kita sebagai programmer bisa betulin. Bayangkan kalau errornya malah muncul di toast (semisal kita hubungkan ke toast).

Simple rule:
Catch gets all errors.
In the catch (err) {...} block we analyze the error object err.
If we don’t know how to handle it, we do throw err.

Reply:
Tidak perlu, jika kita tidak tahu cara untuk menghandle error tersebut, kita harus throw kembali errornya, ini dinamakan practice untuk `avoid silent failures`.

Contohnya, programmer melakukan typo seperti ini:

```
funcUtilize(parsdeData)
```
Hal seperti itu daripada di handle justru lebih baik kalau menghentikan eksekusi supaya kita sebagai programmer bisa betulin. Bayangkan kalau errornya malah muncul di toast (semisal kita hubungkan ke toast).

Simple rule:
- Catch gets all errors.
- In the catch (err) {...} block we analyze the error object err.
- If we don’t know how to handle it, we do throw err.

Reply:
Ok, berarti seperti codingan saya yang terakhir ini apakah sudah benar?

Reply:
Yes. Sudah benar pak.